src/pixiewps.c

```diff
@@ -65,6 +67,89 @@ static const struct option long_options[] = {
 	{  0,          0,                 0,  0  }
 };
 
+#define SECS_PER_JOB_BLOCK 1000
+
+struct crack_job {
+	pthread_t thr;
+	time_t start;
+};
+
+static struct job_control {
+	int jobs;
+	time_t end;
+	uint32_t randr_enonce[4];
+	struct crack_job *crack_jobs;
+	volatile uint32_t print_seed;
+} job_control;
+
+static void* crack_thread(void *arg) {
+	struct crack_job * j = arg;
+	struct random_data *buf = calloc(1, sizeof(struct random_data));
+	char *rand_statebuf = calloc(1, 128);
+
+	uint32_t seed = j->start;
+	uint32_t limit = job_control.end;
+	initstate_r(seed, rand_statebuf, 128, buf);
+	int32_t res = 0;
+
+	while (!job_control.print_seed) {
+		srandom_r(seed, buf);
+		unsigned i;
+		for (i = 0; i < 4; i++) {
+			random_r(buf, &res);
+			if ((uint32_t) res != job_control.randr_enonce[i])
+				break;
+		}
+
+		if (i == 4) {
+			job_control.print_seed = seed;
+			DEBUG_PRINT("Seed found");
+		}
+
+		seed--;
+		if(seed < j->start - SECS_PER_JOB_BLOCK) {
+			j->start -= SECS_PER_JOB_BLOCK * job_control.jobs;
+			seed = j->start;
+			if (seed < limit) break;
+		}
+	}
+	return 0;
+}
+
+static void init_crack_jobs(int jobs, struct global *wps) {
+	job_control.jobs = jobs;
+	job_control.end = wps->end;
+	job_control.print_seed = 0;
+	memset(job_control.randr_enonce, 0, sizeof(job_control.randr_enonce));
+	/* Converting enrollee nonce to the sequence may be generated by current random function */
+	int i, j = 0;
+	for (int i = 0; i < 4; i++) {
+		job_control.randr_enonce[i] |= wps->e_nonce[j++];
+		job_control.randr_enonce[i] <<= 8;
+		job_control.randr_enonce[i] |= wps->e_nonce[j++];
+		job_control.randr_enonce[i] <<= 8;
+		job_control.randr_enonce[i] |= wps->e_nonce[j++];
+		job_control.randr_enonce[i] <<= 8;
+		job_control.randr_enonce[i] |= wps->e_nonce[j++];
+	}
+	job_control.crack_jobs = malloc(jobs * sizeof (struct job_control));
+	time_t curr = wps->start;
+	for(i=0; i<jobs; i++) {
+		job_control.crack_jobs[i].start = curr;
+		pthread_create(&job_control.crack_jobs[i].thr, NULL, crack_thread, &job_control.crack_jobs[i]);
```

Comment:
it's advisable in the long term to initialize a pthread_attr_t here and set its stack size attribute to some small value, like 16 KB, so there's no memory waste going on (GLIBC defaults to insanely huge thread stacks of 8MB, so by using 8 threads one would waste 64MB for basically no gain), since the crack threads effectively uses less than 100 bytes of stack, everything else is dynamically allocated.